pkg/logs/decoder/auto_multiline_handler.go

```diff
@@ -0,0 +1,221 @@
+	timeoutTimer      *time.Timer
+	detectedPattern   *DetectedPattern
+}
+
+// NewAutoMultilineHandler returns a new AutoMultilineHandler.
+func NewAutoMultilineHandler(outputChan chan *Message,
+	lineLimit, linesToAssess int,
+	matchThreshold float64,
+	matchTimeout time.Duration,
+	flushTimeout time.Duration,
+	source *config.LogSource,
+	additionalPatterns []*regexp.Regexp,
+	detectedPattern *DetectedPattern,
+) *AutoMultilineHandler {
+
+	// Put the user patterns at the beginning of the list so we prioritize them if there is a conflicting match.
+	patterns := append(additionalPatterns, formatsToTry...)
+
+	scoredMatches := make([]*scoredPattern, len(patterns))
+	for i, v := range patterns {
+		scoredMatches[i] = &scoredPattern{
+			score:  0,
+			regexp: v,
+		}
+	}
+	h := &AutoMultilineHandler{
+		inputChan:       make(chan *Message),
+		outputChan:      outputChan,
+		isRunning:       true,
+		lineLimit:       lineLimit,
+		matchThreshold:  matchThreshold,
+		scoredMatches:   scoredMatches,
+		linesToAssess:   linesToAssess,
+		flushTimeout:    flushTimeout,
+		source:          source,
+		timeoutTimer:    time.NewTimer(matchTimeout),
+		detectedPattern: detectedPattern,
+	}
+
+	h.singleLineHandler = NewSingleLineHandler(outputChan, lineLimit)
+	h.processsingFunc = h.processAndTry
+
+	return h
+}
+
+// Handle puts all new lines into a channel for later processing.
+func (h *AutoMultilineHandler) Handle(input *Message) {
+	h.inputChan <- input
+}
+
+// Stop stops the handler.
+func (h *AutoMultilineHandler) Stop() {
+	close(h.inputChan)
+}
+
+// Start starts the handler.
+func (h *AutoMultilineHandler) Start() {
+	go h.run()
+}
+
+// run consumes new lines and processes them.
+func (h *AutoMultilineHandler) run() {
+	for {
+		if !h.isRunning {
+			return
+		}
+		line, isOpen := <-h.inputChan
+		if !isOpen {
+			close(h.outputChan)
+			return
+		}
+		h.processsingFunc(line)
+	}
+}
+
+func (h *AutoMultilineHandler) processAndTry(message *Message) {
+	// Process message before anything else
+	h.singleLineHandler.process(message)
+
+	for i, scoredPattern := range h.scoredMatches {
+		match := scoredPattern.regexp.Match(message.Content)
+		if match {
+			log.Tracef("A regexp matched during multiline auto sensing: %v", scoredPattern.regexp)
+			scoredPattern.score++
+
+			// By keeping the scored matches sorted, the best match always comes first. Since we expect one timestamp to match overwhelmingly
+			// it should match most often causing few re-sorts.
+			if i != 0 {
+				sort.Slice(h.scoredMatches, func(i, j int) bool {
+					return h.scoredMatches[i].score > h.scoredMatches[j].score
+				})
+			}
+			break
+		}
+	}
+
+	timeout := false
+	select {
+	case <-h.timeoutTimer.C:
+		log.Debug("Multiline auto detect timed out before reaching line test threshold")
+		timeout = true
+		break
+	default:
+		break
+	}
+
+	h.linesTested++
+	if h.linesTested >= h.linesToAssess || timeout {
+		topMatch := h.scoredMatches[0]
```

Comment:
Quick check: there is no branch where `scoredMatches` could be empty?

Reply:
👍 Nope. The hardcoded list of patterns is always set to the `scoredMatches` prop, and they are never evicted from the list so this is safe.